### ISSUE TITLE
velocity dependent potentials

### ISSUE THREAD

**Issue by antoniommateo** [trotter-suzuki-mpi/trotter-suzuki-mpi#158]
Is it possible to introduce a velocity dependent "potential" in the Hamiltonian. To be more specific, is it possible to introduce in the Hamiltonian a term (operator) like this:  v_R \partial_x 
where v_R is a constant and \partial_x is the derivative respect the x coordinate,
so that  the dynamics in a reference frame moving with speed v_R could be simulated?

The angular rotation term, which is already implemented, is a particular case of this "potentials". 
Is is possible (or not too difficult to implement) in this case to select different angular rotations for each of the component in the two-component condensates? Or, equivalently, is it possible to choose rotation for one component but not for the other one?

**Comment by Lucacalderaro** [comment]
Sorry for the late reply. The implementation of an operator of the kind v_R \partial_x in the Hamiltonian is possible by adding a function that calculates the following formula
$\ket{\psi'} = e^{-i \Delta t/ \hbar O} \ket{\psi}$ where:

- $\Delta t$ is the time interval of a single step in the evolution.
- $\ket{\psi}$ is the state.
- $O$ is the Hermitian operator in the Hamiltonian to be added.

We work in the coordinate representation, where the lattice is a finite sample of the basis $\ket{x,y}$. Then the calculation above becomes:

$\psi'(x',y') =  \sum_{x',y',x,y} ( \bra{x',y'} e^{-i \Delta t/ \hbar O} \ket{x,y} )  \psi(x,y)$.

So the wave function \psi(x,y) is changed into \psi'(x,y) and one needs to be able to calculate the coordinate representation of the operator to be added.

We call these type of functions `block_kernel_something`, for instance the function `block_kernel_rotation` calculate the above formula for the angular momentum operator (O = L_z). These functions are then called in `full_step` and `full_step_imaginary` to build the evolution operator for real and imaginary evolution.

Basically, the function that you want to add is already present in `block_kernel_rotation`: the first three nested for loops calculate the above formula for O = \alpha_y Y P_x, where \alpha_y is a constant, Y is the position operator and P_x is the momentum operator. You can get rid of Y substituting the line 
`double alpha_yy = - 0.5 * alpha_y * y;`
with
`double alpha_yy = - 0.5 * alpha_y;`.
In this way, you are left with the operator O = \alpha_y P_x which is what you want.

To summarize you should create a new function, say `block_kernel_shift_x`, make `full_step` call it and make the constant v_R a property of the Hamiltonian class.

Regarding the second question, do you mean two `angular_velocity` one for each component?

**Comment by antoniommateo** [comment]
Hi Luca,
Thank you for your detailed answer, and I am very sorry for my very late
reply.

The process is now clear to me, and it will be just a question of time to
implement it.

And yes, the second question was meant to be about two angular_velocity,
one for each component.
Thanks.
